google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -1485,129 +1956,204 @@ PooledSession getReadSession() throws SpannerException {
-          }
-        } catch (Throwable t) {
-          SpannerException e = newSpannerException(t);
-          if (!isClosed()) {
-            handlePrepareSessionFailure(e, sess, false);
+          inProcessPrepare = true;
+          numSessionsInProcessPrepared++;
+          PooledSession readSession = readSessions.poll();
+          if (readSession != null) {
+            // Create a read/write transaction in-process if there is already a queue for prepared
+            // sessions. This is more efficient than doing it asynchronously, as it scales with
+            // the number of user threads. The thread pool for asynchronously preparing sessions
+            // is fixed.
+            span.addAnnotation(
+                "Acquired read only session. Preparing in-process for read write transaction");
+            sess = readSession;
+          } else {
+            span.addAnnotation("No session available");
+            maybeCreateSession();
           }
-          sess = null;
-          if (!isSessionNotFound(e)) {
-            throw e;
+        }
+        if (sess == null) {
+          waiter = new WaiterFuture();
+          if (inProcessPrepare) {
+            // inProcessPrepare=true means that we have already determined that the queue for
+            // preparing read/write sessions is larger than the number of threads in the prepare
+            // thread pool, and that it's more efficient to do the prepare in-process. We will
+            // therefore create a waiter for a read-only session, even though a read/write session
+            // has been requested.
+            readWaiters.add(waiter);
+          } else {
+            readWriteWaiters.add(waiter);
           }
         }
+      } else {
+        span.addAnnotation("Acquired read write session");
       }
-      if (sess != null) {
-        break;
-      }
+      return checkoutSession(span, sess, waiter, true, inProcessPrepare);
     }
-    sess.markBusy();
-    incrementNumSessionsInUse();
-    span.addAnnotation(sessionAnnotation(sess));
-    return sess;
   }
 
-  PooledSession replaceReadSession(SessionNotFoundException e, PooledSession session) {
+  private PooledSessionFuture checkoutSession(
+      final Span span,
+      final PooledSession readySession,
+      WaiterFuture waiter,
+      boolean write,
+      final boolean inProcessPrepare) {
+    Future<PooledSession> sessionFuture;
+    if (waiter != null) {
+      logger.log(
+          Level.FINE,
+          "No session available in the pool. Blocking for one to become available/created");
+      span.addAnnotation(
+          String.format(
+              "Waiting for %s session to be available", write ? "read write" : "read only"));
+      sessionFuture = waiter;
+    } else {
+      sessionFuture = ApiFutures.immediateFuture(readySession);
+    }
+    SimpleForwardingFuture<PooledSession> forwardingFuture =
+        new SimpleForwardingFuture<SessionPool.PooledSession>(sessionFuture) {
```

Comment:
This is a really long anonymous class implementation. Would it make sense to externalize it to a top-level class or at least least a nested class to improve readability of `checkoutSession()`?

Reply:
I've made a nested class and added some documentation. Externalizing it as a top-level class is less logical, as it depends heavily on the `SessionPool` class.